File: 2023/day14.py
import os
import sys
import itertools

test = """\
O....#....
O.OO#....#
.....##...
OO.#O....O
.O.....O#.
O.#..O.#.#
..O..#O..O
.......O..
#....###..
#OO..#...."""

day = os.path.splitext(os.path.basename(__file__))[0]

if 'test' in sys.argv:
    data = test
else:
    data = open(day+'.txt').read()


DEBUG = 'debug' in sys.argv

# Convert to a dict.

round = []
cube = set()
for y,row in enumerate(data.splitlines()):
    for x,c in enumerate(row):
        if c == '#':
            cube.add( (x,y) )
        elif c == 'O':
            round.append( (x,y) )

H = y+1
W = x+1

def tilt_n(round):
    round = sorted(round, key=lambda z: z[1])
    newround = []
    for x,y in round:
        while y and (x,y-1) not in newround and (x,y-1) not in cube:
            y -= 1
        newround.append( (x,y) )
    return newround

def tilt_w(round):
    # Sort by X.
    round = sorted(round)
    newround = []
    for x,y in round:
        while x and (x-1,y) not in newround and (x-1,y) not in cube:
            x -= 1
        newround.append( (x,y) )
    return newround

def tilt_s(round):
    # Sort by Y reversed.
    round = sorted(round, key=lambda z: -z[1])
    newround = []
    for x,y in round:
        while y < H-1 and (x,y+1) not in newround and (x,y+1) not in cube:
            y += 1
        newround.append( (x,y) )
    return newround

def tilt_e(round):
    # Sort by X reversed.
    round = sorted(round, reverse=True)
    newround = []
    for x,y in round:
        while x < W-1 and (x+1,y) not in newround and (x+1,y) not in cube:
            x += 1
        newround.append( (x,y) )
    return newround

def weight(round):
    return sum(H-z[1] for z in round)

def part1(round):
    return weight(tilt_n(round))

def part2(round):
    seen = [0]
    scores = [0]
    want = -1
    for i in itertools.count():
        round = tilt_n(round)
        round = tilt_w(round)
        round = tilt_s(round)
        round = tilt_e(round)
        cur = hash(tuple(round))
        scores.append(weight(round))
        if DEBUG:
            print(i+1,scores[-1])
        if cur in seen:
            pat0 = seen.index(cur)
            cycle = len(seen) - pat0
            want =  (1000000000 - pat0) % cycle + pat0
            break
        seen.append(cur)
    return scores[want]


print("Part 1:", part1(round)) # 64
print("Part 2:", part2(round)) # 83516

File: 2023/day14.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <cstring>
#include <vector>
#include <map>
#include <deque>
#include <tuple>
#include <algorithm>
#include <numeric>

using namespace std;

bool DEBUG = false;
bool TEST = false;

typedef vector<string> StringVector;
typedef vector<int> IntVector;
typedef vector<int64_t> LongVector;

const string test(
"O....#....\n"
"O.OO#....#\n"
".....##...\n"
"OO.#O....O\n"
".O.....O#.\n"
"O.#..O.#.#\n"
"..O..#O..O\n"
".......O..\n"
"#....###..\n"
"#OO..#...."
);

int W = 0;
int H = 0;

// Parse the input.

void parse( istream & fin, StringVector & lines )
{
    string line;
    while( getline( fin, line ) )
        lines.push_back(line);
}

void tilt_n( StringVector & grid )
{
    for( int x = 0; x < W; x++)
    {
        int dy = 0;
        for( int y = 0; y < H; y++ )
        {
            char c = grid[y][x];
            if( c == '.' )
                dy++;
            else if( c == '#')
                dy = 0;
            else if( c == 'O')
            {
                grid[y][x] = '.';
                grid[y-dy][x] = 'O';
            }
        }
    }
}

void tilt_s( StringVector & grid )
{
    for( int x = 0; x < W; x++)
    {
        int dy = 0;
        for( int y = H-1; y >= 0; y-- )
        {
            char c = grid[y][x];
            if( c == '.' )
                dy++;
            else if( c == '#')
                dy = 0;
            else if( c == 'O')
            {
                grid[y][x] = '.';
                grid[y+dy][x] = 'O';
            }
        }
    }
}


void tilt_w( StringVector & grid )
{
    for( int y = 0; y < H; y++ )
    {
        int dx = 0;
        for( int x = 0; x < W; x++)
        {
            char c = grid[y][x];
            if( c == '.' )
                dx++;
            else if( c == '#')
                dx = 0;
            else if( c == 'O')
            {
                grid[y][x] = '.';
                grid[y][x-dx] = 'O';
            }
        }
    }
}


void tilt_e( StringVector & grid )
{
    for( int y = 0; y < H; y++ )
    {
        int dx = 0;
        for( int x = W-1; x >= 0; x--)
        {
            char c = grid[y][x];
            if( c == '.' )
                dx++;
            else if( c == '#')
                dx = 0;
            else if( c == 'O')
            {
                grid[y][x] = '.';
                grid[y][x+dx] = 'O';
            }
        }
    }
}


int weight( StringVector & grid )
{
    int sum = 0;
    for( int y = 0; y < H; y++ )
        sum += (H-y)*count(grid[y].begin(),grid[y].end(),'O');
    return sum;
}


uint64_t unique( StringVector & grid )
{
    // Join into a single string.
    string mass;
    for( auto & s : grid )
        mass += s;
    return hash<string>{}(mass);
}


void print( StringVector & grid )
{
    cout << "----\n";
    for( auto & s : grid )
        cout << s << "\n";
}


int part1( StringVector grid )
{
    tilt_n(grid);
    return weight(grid);
}


int part2( StringVector grid )
{
    map<uint64_t,int> seen;
    seen[0] = 0;
    IntVector scores(1);
    int pat0 = 0;
    for( ;; )
    {
        tilt_n(grid);
        tilt_w(grid);
        tilt_s(grid);
        tilt_e(grid);
        uint64_t cur = unique(grid);
        scores.push_back(weight(grid));
        if( DEBUG )
            cout << seen.size() << " " << scores.back() << " " << cur << "\n";
        if( seen.find(cur) != seen.end() )
        {
            pat0 = seen[cur];
            break;
        }
        seen[cur] = seen.size();
    }
    int cycle = seen.size() - pat0;
    int want =  (1000000000 - pat0) % cycle + pat0;
    return scores[want];
}


int main( int argc, char ** argv )
{
    while( *++argv )
    {
        string arg(*argv);
        if( arg == "debug")
            DEBUG = true;
        else if( arg =="test")
            TEST = true;
    }

    StringVector lines;
    if( TEST )
    {
        istringstream data;
        data.str(test);
        parse(data, lines);
    }
    else 
    {
        ifstream data("day14.txt");
        parse(data, lines);
    }

    W = lines[0].size();
    H = lines.size();

    cout << "Part 1: " << part1(lines) << "\n";
    cout << "Part 2: " << part2(lines) << "\n";
}
